### PR TITLE
Visualise low-dimensional embeddings in spiral classification notebook

### PR DESCRIPTION
**Pull request by jbfiot**
This PR adds simple code to visualise and play with 2D embeddings in the spiral classifcation notebook. 

Output example:

Screenshot from 2022-05-15 10-39-56

### CODE REVIEW

File: 04-spiral_classification.ipynb [hunk 1]
```diff
@@ -308,6 +308,78 @@
...
+    "optimiser = torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=lambda_l2)\n",
+    "\n",
+    "for t in range(2_000):\n",
+    "    l = model(X)\n",
+    "    F = C(l, y)\n",
+    "    L = F.mean()\n",
+    "    optimiser.zero_grad()\n",
+    "    L.backward()\n",
+    "    optimiser.step()\n",
+    "    overwrite(f'[EPOCH]: {t}, [LOSS]: {L.item():.6f}, [ACCURACY]: {acc(l, y):.3f}')\n"
```
Comment (Atcold) [inline 04-spiral_classification.ipynb:343]: Why is there another training here?

Comment (jbfiot) [reply 04-spiral_classification.ipynb:343]: I re-used the variable name `model`, though it is a different model with an extra linear layer, such that data goes $\mathbb{R}^n \rightarrow \mathbb{R}^d \rightarrow \mathbb{R}^2 \rightarrow \mathbb{R}^K$ instead of $\mathbb{R}^n \rightarrow \mathbb{R}^d \rightarrow \mathbb{R}^K$ in the previous one. 

The added code just trains that new model and plots the internal 2D embeddings from that extra layer. 

Please feel free to close the pull request without merging it if this adds too much code and not much value ^^

File: 04-spiral_classification.ipynb [hunk 2]
```diff
@@ -308,6 +308,78 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "%matplotlib inline\n",
+    "set_default()\n",
```
Comment (Atcold) [inline 04-spiral_classification.ipynb:373]: Do we need these?

File: 04-spiral_classification.ipynb [hunk 3]
```diff
@@ -308,6 +308,78 @@
...
+    "\n",
+    "def get_layer_outputs(name):\n",
+    "    def hook(model, input, output):\n",
+    "        layer_outputs[name] = output.detach()\n",
+    "    return hook\n",
+    "\n",
+    "model[-2].register_forward_hook(get_layer_outputs('low_dim_embeddings'))\n",
+    "\n",
+    "# pass data through model to populate layer_outputs\n",
+    "output = model(X)"
```
Comment (Atcold) [inline 04-spiral_classification.ipynb:363]: Let's hide this in the visualisation library?
It's rather 'advanced' stuff and this is the first notebook of the course, haha.

File: 04-spiral_classification.ipynb (hunk 1 above)
Comment (Atcold) [reply 04-spiral_classification.ipynb:343]: OMG, $\LaTeX$ on GitHub! I saw the announcement, but I haven't used it yet! Aww! ❤️

I fear that this would be take away the attention from the major points.
There's the linear model.
There's the 'deep model'.
There are two ways of looking at the energy.
This notebook doesn't focus on classification.
Classification is what we use to introduce a bunch of other things.

I don't want to close this PR.
Let me think what we could do.

File: 04-spiral_classification.ipynb (hunk 2 above)
Comment (jbfiot) [reply 04-spiral_classification.ipynb]: Previously yes (at least with my local config, as the code was after `%matplotlib notebook`). 
This is now removed in the PR update :)

File: 04-spiral_classification.ipynb (hunk 3 above)
Comment (jbfiot) [reply 04-spiral_classification.ipynb]: Good idea! Done in the PR update :)

**Review by Atcold — APPROVED** [review]
I like it. Thanks.

**Comment by Atcold** [comment]
Oh, LOL, we forgot to add the linear decision boundaries! 😅

**Comment by Atcold** [comment]
The code would be this one.

```python
if layer.__class__ == torch.nn.modules.linear.Linear and layer.out_features == 2:
        layer.register_forward_hook(get_layer_outputs('low_dim_embeddings'))
        model(X)  # pass data through model to populate layer_outputs
        plot_data(layer_outputs['low_dim_embeddings'], y, label='e', zoom=10)
        last_layer = model[-1]
        mesh = torch.arange(-1.1, 1.1, 0.01) * 10
        xx, yy = torch.meshgrid(mesh, mesh, indexing='ij')
        with torch.no_grad():
            data = torch.stack((xx.reshape(-1), yy.reshape(-1)), dim=1)
            Z = last_layer(data)
        Z = Z.argmax(dim=1).reshape(xx.shape)
        contourf(xx, yy, Z, cmap=cm.Spectral, alpha=0.3, levels=y.max().item())
        return layer_outputs['low_dim_embeddings']
```

Also, please, don't use `detach()`. Rather, use the context manager.

**Comment by Atcold** [comment]
Also, this is what I'm currently putting together for the book :)

image

**Comment by Atcold** [comment]
More context [here](https://twitter.com/alfcnz/status/1534403904267591683).

**Comment by jbfiot** [comment]
> The code would be this one.
> 
> ```python
> if layer.__class__ == torch.nn.modules.linear.Linear and layer.out_features == 2:
>         layer.register_forward_hook(get_layer_outputs('low_dim_embeddings'))
>         model(X)  # pass data through model to populate layer_outputs
>         plot_data(layer_outputs['low_dim_embeddings'], y, label='e', zoom=10)
>         last_layer = model[-1]
>         mesh = torch.arange(-1.1, 1.1, 0.01) * 10
>         xx, yy = torch.meshgrid(mesh, mesh, indexing='ij')
>         with torch.no_grad():
>             data = torch.stack((xx.reshape(-1), yy.reshape(-1)), dim=1)
>             Z = last_layer(data)
>         Z = Z.argmax(dim=1).reshape(xx.shape)
>         contourf(xx, yy, Z, cmap=cm.Spectral, alpha=0.3, levels=y.max().item())
>         return layer_outputs['low_dim_embeddings']
> ```
> 
> Also, please, don't use `detach()`. Rather, use the context manager.

Thanks! Created #34 :)

**Comment by jbfiot** [comment]
> Also, this is what I'm currently putting together for the book :) Can you tell what it is?
> 
> image

Cool plot! 

I imagine the points represent the original observations in $\mathbb{R}^2$, and the arrows illustrate how these points are "moved" towards the "new" points in the $\mathbb{R}^2$ embeddings space?

**Comment by jbfiot** [comment]
> More context [here](https://twitter.com/alfcnz/status/1534403904267591683).

Awesome! Looking forward to the early draft you mentioned on Twitter!
